src/shared/dissect-image.c

```diff
@@ -1255,12 +1256,27 @@ int dissect_image_file(
 }
 
 DissectedImage* dissected_image_unref(DissectedImage *m) {
+        int r;
+
         if (!m)
                 return NULL;
 
         /* First, clear dissected partitions. */
-        for (PartitionDesignator i = 0; i < _PARTITION_DESIGNATOR_MAX; i++)
-                dissected_partition_done(m->partitions + i);
+        for (PartitionDesignator i = 0; i < _PARTITION_DESIGNATOR_MAX; i++) {
+                DissectedPartition *p = m->partitions + i;
+
+                if (p->mount_node_fd >= 0) {
+                        /* Let's make sure the mount node fd is closed before we tell btrfs to forget the
+                         * device. */
+                        p->mount_node_fd = safe_close(p->mount_node_fd);
+
+                        r = btrfs_forget_device(p->node);
+                        if (r < 0)
+                                log_debug_errno(r, "Failed to forget btrfs device %s, ignoring: %m", p->node);
+                }
```

Comment:
hmm, so the other stuff we do in dissected_image_unref() is not destructive per se (only if the ref count hits zero and the device wasn't "relinquished" yet at that time).

i think we probably shouldn't do this here hence, we shouldn't do destructive stuff in a destructor that some code wants to use entirely "passively". (for example gpt-auto-generator uses all this passively, it doesn't take possession of anything, it just uses it to generate units for later use of the device).

so where to do this instead? I think at two places:

1. inside of decrypted_image_free() right before detaching encrypted volumes
2. inside loop_device_free() right before block_device_remove_all_partitions()

or in other words: right where we are about to destroy the partition block devices.

Reply:
I just moved it into `block_device_remove_all_partitions()` as that seemed like the better place to put it